pkg/collector/providers/docker.go

```diff
@@ -35,27 +40,71 @@ func (d *DockerConfigProvider) String() string {
 }
 
 // Collect retrieves all running containers and extract AD templates from their labels.
-// TODO: suscribe to docker events and only invalidate cache if we get a `start` event since last Collect.
 func (d *DockerConfigProvider) Collect() ([]check.Config, error) {
 	var err error
 	if d.dockerUtil == nil {
 		d.dockerUtil, err = docker.GetDockerUtil()
 		if err != nil {
 			return []check.Config{}, err
 		}
+		go d.listen()
 	}
 
 	containers, err := d.dockerUtil.AllContainerLabels()
 	if err != nil {
 		return []check.Config{}, err
 	}
 
+	d.Lock()
+	d.upToDate = true
+	d.Unlock()
+
 	return parseDockerLabels(containers)
 }
 
-// IsUpToDate updates the list of AD templates versions in the Agent's cache and checks the list is up to date compared to Docker's data.
+// We listen to docker events and invalidate our cache when we receive a start/die event
+func (d *DockerConfigProvider) listen() {
+	d.Lock()
+	d.streaming = true
+	d.Unlock()
+
+CONNECT:
+	for {
+		eventChan, errChan, err := d.dockerUtil.SubscribeToContainerEvents(d.String())
+		if err != nil {
+			log.Warnf("error subscribing to docker events: %s", err)
+			break CONNECT // We disable streaming and revert to always-pull behaviour
+		}
+
+		for {
+			select {
+			case ev := <-eventChan:
+				if ev.Action == "start" || ev.Action == "die" {
```

Comment:
why only using start and die ? we should also flush the cache if there is a create, destroy, kill, oom, rename, resize, restart, stop, unpause. I feel like most of the events would have stop in their flow (like oom I assume would be followed or preceded by a stop), but not sure about event actions like rename, resize or pause.

Reply:
- rename, resize or pause will not modify container labels, so that means we don't need to look for other templates

But maybe `create` and `destroy` are more appropriate than `start`/`die`. As containers labels can't be changed once it is created, we're covered.

Reply:
@CharlyF after taking another look at it, we parse the running container list (`docker ps`), so `start` and `die` are more appropriate than `create` and `destroy`: if a container is created but started only 30 seconds later we will miss it: we will `Collect` without it (it is not running yet) and will not invalidate the cache after.

I added a comment to explain it to whoever will have to 🔍 it later ;)